django_mongodb_cli/repo.py

```diff
@@ -0,0 +1,472 @@
+import os
+import subprocess
+
+import click
+from .config import test_settings_map
+from .utils import (
+    apply_patches,
+    copy_mongo_apps,
+    copy_mongo_migrations,
+    copy_mongo_settings,
+    get_management_command,
+    get_repos,
+    repo_clone,
+    repo_fetch,
+    repo_install,
+    repo_status,
+    repo_update,
+)
+
+
+class Repo:
+    def __init__(self):
+        self.home = "src"
+        self.config = {}
+
+    def set_config(self, key, value):
+        self.config[key] = value
+
+    def __repr__(self):
+        return f"<Repo {self.home}>"
+
+
+pass_repo = click.make_pass_decorator(Repo)
+
+
+@click.group(invoke_without_command=True)
+@click.option(
+    "-l",
+    "--list-repos",
+    is_flag=True,
+    help="List all repositories in `pyproject.toml`.",
+)
+@click.pass_context
+def repo(ctx, list_repos):
+    """
+    Run Django fork and third-party library tests.
+    """
+    ctx.obj = Repo()
+    repos, url_pattern, branch_pattern, upstream_pattern = get_repos("pyproject.toml")
+    if list_repos:
+        for repo_entry in repos:
+            click.echo(repo_entry)
+        return
+    if ctx.invoked_subcommand is None:
+        click.echo(ctx.get_help())
+
+
+@repo.command()
+@click.argument("repo_names", nargs=-1, required=False)
+@click.option(
+    "-a",
+    "--all-repos",
+    is_flag=True,
+)
+@click.option(
+    "-i",
+    "--install",
+    is_flag=True,
+)
+@click.pass_context
+@pass_repo
+def clone(repo, ctx, repo_names, all_repos, install):
+    """Clone repositories from `pyproject.toml`."""
+    repos, url_pattern, branch_pattern, upstream_pattern = get_repos("pyproject.toml")
+
+    if repo_names:
+        for repo_name in repo_names:
+            not_found = set()
+            for repo_entry in repos:
+                if (
+                    os.path.basename(url_pattern.search(repo_entry).group(0))
+                    == repo_name
+                ):
+                    repo_clone(repo_entry, url_pattern, branch_pattern, repo)
+                    if install:
+                        clone_path = os.path.join(ctx.obj.home, repo_name)
+                        if os.path.exists(clone_path):
+                            repo_install(clone_path)
+                    return
+                else:
+                    not_found.add(repo_name)
+            click.echo(f"Repository '{not_found.pop()}' not found.")
+        return
+
+    if all_repos:
+        click.echo(f"Cloning {len(repos)} repositories...")
+        for repo_entry in repos:
+            repo_clone(repo_entry, url_pattern, branch_pattern, repo)
+        return
+
+    if ctx.args == []:
+        click.echo(ctx.get_help())
```

Comment:
Couldn't pip install from pyproject be the proxy for installing these?
something like pip install ".[django_rest_framework]"?

Reply:
That syntax is for [optional dependencies](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies) which I don't think can be "editable" or downloadable via `git+ssh`. However, it's possible this could be improved by having `pyproject.toml` read and install packages from `requirements.txt` which can contain editable `git+ssh` package references.